### PR TITLE
Data apps model actions implicit creator

### PR DESCRIPTION
**Pull request by iethree**
## Description

This branch focuses on enabling implicit actions on models, and has some nice side-effects in terms of unifying how we treat implicit and explicit actions. Namely, on the backend, there is now a data object for implicit actions, it just doesn't have a solidified query, so the backend dynamically generates the necessary parameters for `insert`, `update`, and `delete` actions.

- add UI for enabling implicit actions for a given model

enableImplicitActions

- link buttons or forms to implicit actions, just like explicit actions

linkimplicitactiontobutton

- disables the edit and delete buttons on the list view for now, until we figure out what to do with those

### CODE REVIEW

**Comment by kulyk** [comment]
Would be cool to have an "Create new action" button available on the model detail page after implicit actions are created

<details>
<summary>Without actions</summary>

no-actions-model

</details>

<details>
<summary>With implicit actions</summary>

model-with-implicit-actions

</details>

**Comment by kulyk** [comment]
@iethree noticed a few things while playing around with that branch. I'm not sure if that could be me doing something wrong with branch setup, would appreciate if you can take a look

https://user-images.githubusercontent.com/17258145/194920366-56415b78-2a80-4b3f-91f0-b1855cba21d4.mp4

**Comment by iethree** [comment]
> @iethree noticed a few things while playing around with that branch. I'm not sure if that could be me doing something wrong with branch setup, would appreciate if you can take a look
> 
>  CleanShot.2022-10-10.at.18.09.20.mp4

all good observations.
- the problems with detail view appear to be:
  - we're losing the query param in the link in list view
  - the scaffolding link between the model detail view and the filter is broken
- the extra filters come automatically from having an embedded model instead a question
- I share your sadness about the step backwards for forms, but I think it's important to unify these things from the backend parameters in a way that works for all actions, and separately add logic to fetch field metadata to get back to the UX we had before for implicit forms
- the weirdness with the delete button is that the backend labels the button "delete", but the model doesn't have any implicit actions yet (so when you connected the insert action, it worked, but the button was still labeled 'delete'). Case is working on auto-creating implicit actions as part of scaffolding, so this won't happen any more. This also gets much better when we no longer add/change actions from the click actions panel

**Comment by kulyk** [comment]
@iethree gotcha, thanks for explaining 🙏

File: frontend/src/metabase/entities/actions/actions.ts [hunk 1]
```diff
@@ -71,6 +76,29 @@ const associateAction = ({
...
+  delete: true,
+};
+
+const enableImplicitActionsForModel =
+  async (modelId: number, options = defaultImplicitActionCreateOptions) =>
+  async (dispatch: Dispatch) => {
+    await Promise.all(
+      Object.entries(options)
+        .filter(([, shouldCreate]) => !!shouldCreate)
+        .map(([method]) =>
```
Comment (kulyk) [inline frontend/src/metabase/entities/actions/actions.ts:91]: I personally find this a bit hard to follow, had to pause at this block for a minute to parse what exactly is happening here

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx [hunk 2]
```diff
@@ -8,26 +10,42 @@ import Icon from "metabase/components/Icon";
...
   if (!actions?.length) {
     return (
       <EmptyStateContainer>
         <EmptyStateTitle>{t`This model does not have any actions yet.`}</EmptyStateTitle>
+        <Button onClick={createImplicitActions} icon="add">
+          {t`Enable implicit actions`}
+        </Button>
         <Button
           as={Link}
           to={`/action/create`}
```
Comment (kulyk) [inline frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx:51]: nit (not really related to this PR):

1. Usually it's preferred not to use backticks when you don't need a template string (it can be confusing)
2. For strings I guess it's preferred to use a React string prop shortcut like `to="/action/create"`

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx [hunk 3]
```diff
@@ -8,26 +10,42 @@ import Icon from "metabase/components/Icon";
...
 
-function ModelActionDetails({ actions }: Props) {
+function ModelActionDetails({
+  actions,
+  modelId,
+  enableImplicitActionsForModel,
+}: Props) {
+  const createImplicitActions = async () => {
+    await enableImplicitActionsForModel(modelId);
+  };
```
Comment (kulyk) [inline frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx:40]: I guess it's more common here to use names `handleCreateImplicitActions` kind of naming for callbacks inside the component

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx [hunk 4]
```diff
@@ -37,22 +55,41 @@ function ModelActionDetails({ actions }: Props) {
...
+export default _.compose(
+  Actions.loadList(
+    {
+      query: (state: State, props: { modelId?: number | null }) => ({
+        "model-id": props?.modelId,
+      }),
+    },
+    connect(null, mapDispatchToProps),
+  )(ModelActionDetails),
+);
```
Comment (kulyk) [inline frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx:95]: The `compose` here doesn't seem to do anything, mind breaking the inner part down?

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/utils.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import type { WritebackAction } from "metabase-types/api";
+
+export const hasImplicitActions = (actions: WritebackAction[]): boolean =>
+  actions.some(isImplicitAction);
+
+export const isImplicitAction = (action: WritebackAction): boolean =>
+  action.type === "implicit";
```
Comment (kulyk) [inline frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/utils.ts:7]: Should be probably a part of `metabase/writeback/utils`?

File: frontend/src/metabase/writeback/components/ActionViz/ActionForm.tsx [hunk 6]
```diff
@@ -21,7 +21,7 @@ interface ActionFormProps {
 function ActionForm({ onSubmit, missingParameters, action }: ActionFormProps) {
   return (
     <DataAppPageFormWrapper>
-      <DataAppPageFormTitle>{action.name}</DataAppPageFormTitle>
+      <DataAppPageFormTitle>{action.name ?? action.slug}</DataAppPageFormTitle>
```
Comment (kulyk) [inline frontend/src/metabase/writeback/components/ActionViz/ActionForm.tsx:24]: Question: seems like we need to do this `name || slug` twist in a few places now. Is it possible to generalize this? Maybe assign a name [the same way we assign IDs](https://github.com/metabase/metabase/blob/038b432275a99219f1bf11a8c8bc44795569beca/frontend/src/metabase/entities/actions/actions.ts#L124) now in the Actions entity?

**Review by kulyk — APPROVED** [review]
Looks good overall. Left a few nits and a couple of questions to address first, but don't want to block the merging 👍

File: frontend/src/metabase/entities/actions/actions.ts (hunk 1 above)
Comment (iethree) [reply frontend/src/metabase/entities/actions/actions.ts:91]: very fair. took me a second to figure out what this did too. I've cleared it up a bit.

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx (hunk 2 above)
Comment (iethree) [reply frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionDetails.tsx:51]: 🤔  this feels like a job for the linter. fixed it here manually.

File: frontend/src/metabase/writeback/components/ActionViz/ActionForm.tsx (hunk 6 above)
Comment (iethree) [reply frontend/src/metabase/writeback/components/ActionViz/ActionForm.tsx:24]: that's a fantastic idea! ❤️
